cpp/src/sampling/uniform_neighbor_sampling_impl.hpp

```diff
@@ -149,4 +301,56 @@ uniform_nbr_sample(raft::handle_t const& handle,
     handle, graph_view, edge_weight_view, d_start_vs, fan_out, with_replacement, seed);
 }
 
+template <typename vertex_t,
+          typename edge_t,
+          typename weight_t,
+          typename edge_type_t,
+          bool store_transposed,
+          bool multi_gpu>
+std::tuple<rmm::device_uvector<vertex_t>,
+           rmm::device_uvector<vertex_t>,
+           rmm::device_uvector<edge_t>,
+           std::optional<rmm::device_uvector<edge_type_t>>,
+           std::optional<rmm::device_uvector<weight_t>>,
+           rmm::device_uvector<int32_t>,
+           std::optional<rmm::device_uvector<int32_t>>>
+uniform_neighbor_sample(
+  raft::handle_t const& handle,
+  graph_view_t<vertex_t, edge_t, store_transposed, multi_gpu> const& graph_view,
+  std::optional<edge_property_view_t<edge_t, weight_t const*>> edge_weight_view,
+  std::optional<
+    edge_property_view_t<edge_t,
+                         thrust::zip_iterator<thrust::tuple<edge_t const*, edge_type_t const*>>>>
+    edge_type_view,
+  raft::device_span<vertex_t const> starting_vertices,
+  std::optional<raft::device_span<int32_t const>> starting_labels,
+  raft::host_span<int const> fan_out,
+  bool with_replacement,
+  uint64_t seed)
+{
+  rmm::device_uvector<vertex_t> d_start_vs(starting_vertices.size(), handle.get_stream());
+  raft::copy(
+    d_start_vs.data(), starting_vertices.data(), starting_vertices.size(), handle.get_stream());
+
+  std::optional<rmm::device_uvector<int32_t>> d_start_labels{std::nullopt};
+  if (starting_labels) {
+    d_start_labels = std::make_optional(
+      rmm::device_uvector<int32_t>(starting_labels->size(), handle.get_stream()));
+    raft::copy(d_start_labels->data(),
+               starting_labels->data(),
+               starting_labels->size(),
+               handle.get_stream());
+  }
```

Comment:
I guess this should be unnecessary if the function below takes span for starting vertices and labels.

Reply:
The shuffle functions expect `d_start` and `d_in` to be `rmm::device_uvector` so that we can move them and overwrite them as we execute.

Reply:
So, if I am not mistaken,

If the intention is to move them and overwrite, shouldn't we better pass `d_start_labels` as an R-value? If we're passing `d_start_labels` as an L-value reference, that implies that we're willing to use `d_start_labels` after the function call and `d_start_labels` will be in a certain well-defined state after the function call.

And I guess we don't call the shuffle function in single-GPU, so this overhead can be avoided in signle-GPU, right? Then, shouldn't we better create a copy before calling a shuffle function rather than creating a copy here? And I guess it will make the intention clearer. To groupby elements based on destination GPUs, we need to modify the input, but '`starting_labels` is const, so we need to make a copy. This is clear. But the intention of creating a copy to call a detail space implementation function here is not very clear I think.

Reply:
I modified the API to take R-values, modified the C API and the testers to pass R-values, that allowed me to remove this extra copy.